stats/bridge_stats.go

```diff
@@ -0,0 +1,347 @@
+				log.Warn("Stats login failed", "err", err)
+				conn.Close()
+				errTimer.Reset(10 * time.Second)
+				continue
+			}
+			go s.loop(conn)
+			s.readLoop(conn)
+		}
+	}
+}
+
+func (s *Service) Stop() {
+	close(s.quitCh)
+}
+
+func (s *Service) login(conn *connWrapper) error {
+	// Construct and send the login authentication
+	auth := &authMsg{
+		ID: s.node,
+		Info: &NodeInfo{
+			Node:          s.node,
+			Operator:      s.operator,
+			BridgeVersion: s.version,
+		},
+		Secret: s.pass,
+	}
+	login := map[string][]interface{}{
+		"emit": {"hello", auth},
+	}
+	if err := conn.WriteJSON(login); err != nil {
+		return err
+	}
+	// Retrieve the remote ack or connection termination
+	var ack map[string][]string
+	if err := conn.ReadJSON(&ack); err != nil || len(ack["emit"]) != 1 || ack["emit"][0] != "ready" {
+		return errors.New("unauthorized")
+	}
+	return nil
+}
+
+func (s *Service) report(conn *connWrapper) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	// collect bridge stats and send to ronin stats
+	info := &BridgeInfo{
+		Node:           s.node,
+		Operator:       s.operator,
+		Version:        s.version,
+		LastError:      s.lastError,
+		ProcessedBlock: s.processedBlock,
+	}
+
+	// count pending/failed tasks from database
+	pending, err := s.store.CountTasks(s.chainId, "pending")
+	if err != nil {
+		log.Error("error while getting pending tasks", "err", err)
+	} else {
+		info.PendingTasks = int(pending)
+	}
+
+	failed, err := s.store.CountTasks(s.chainId, "failed")
+	if err != nil {
+		log.Error("error while getting failed tasks", "err", err)
+	} else {
+		info.FailedTasks = int(failed)
+	}
+
+	report := map[string][]interface{}{
+		"emit": {"bridge-stats", info},
+	}
+	return conn.WriteJSON(report)
+}
+
+func (s *Service) loop(conn *connWrapper) {
+	sendStatsTicker := time.NewTicker(10 * time.Second)
+	for {
+		select {
+		case msg := <-s.errCh:
+			s.setLastError(msg.Listener, msg.Err)
+		case msg := <-s.processedBlockCh:
+			s.setProcessedBlock(msg.Listener, msg.ProcessedBlock)
+		case <-sendStatsTicker.C:
+			if err := s.report(conn); err != nil {
+				log.Warn("bridge stats report failed", "err", err)
+			}
+		}
+	}
+}
+
+// readLoop loops as long as the connection is alive and retrieves data packets
+// from the network socket. If any of them match an active request, it forwards
+// it, if they themselves are requests it initiates a reply, and lastly it drops
+// unknown packets.
+func (s *Service) readLoop(conn *connWrapper) {
+	// If the read loop exits, close the connection
+	defer conn.Close()
+
+	for {
+		// Retrieve the next generic network packet and bail out on error
+		var blob json.RawMessage
+		if err := conn.ReadJSON(&blob); err != nil {
+			log.Warn("Failed to retrieve stats server message", "err", err)
+			return
+		}
+		// If the network packet is a system ping, respond to it directly
+		var ping string
+		if err := json.Unmarshal(blob, &ping); err == nil && strings.HasPrefix(ping, "primus::ping::") {
```

Comment:
AFAIK, the bridge pings to stats server and stats server responses with pong. So there is no ping message from stats server.

Reply:
the library itself has its own ping function called system ping. Therefore, we also need to handle pong message